### PR TITLE
iox-#1419 runtime name must be valid file name

### LINKED ISSUE #1419: Slash in runtime names is interpreted as sub path
## Required information

**Observed result or behaviour:**
It is impossible to create a runtime with an app name which has a slash since this is interpreted as path by some underlying constructs.

```cxx
iox::runtime::PoshRuntime::initRuntime("aaargh/aaaaaargh/whoopsie");
```

The problem is caused by multiple posix constructs:
* The `FileLock` which tries to create a file lock under `/tmp/aaargh/aaaaaargh/whoopsie.lock` but the sub folders are not existing
* The `UnixDomainSocket` which tries to create a unix domain socket in `/tmp/aaargh/aaaaaargh/whoopsie`
* The `NamedPipe` which tries to create a shared memory file handle also under such a subpath
* The `NamedSemaphore` in `NamedPipe` with the same issue.

**Solution:**

iceoryx restricts the runtime names to valid file names which the user can verify via `iox::cxx::isValidFileName` from `cxx/helplets.hpp`.

This wouldn't be exactly a public API change since application which non leading slashs in their name would only generate fatal errors in the corresponding posix wrappers when the path does not exist.
But if some smart user would create `mkdir -p /tmp/aaargh/aaaaaargh/` the application would run but since this is a very weird corner case I would dismiss it.

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1419

### CODE REVIEW

File: iceoryx_posh/source/runtime/posh_runtime.cpp [hunk 1]
```diff
@@ -81,18 +82,13 @@ const RuntimeName_t& PoshRuntime::verifyInstanceName(cxx::optional<const Runtime
...
+    else if (!cxx::isValidFileName(**name))
     {
-        LogError() << "Cannot initialize runtime. Application name must not be empty!";
-        errorHandler(PoshError::POSH__RUNTIME_NAME_EMPTY, ErrorLevel::FATAL);
-    }
-    else if (name.value()->c_str()[0] == '/')
-    {
-        LogError() << "Cannot initialize runtime. Please remove leading slash from Application name " << *name.value();
-        errorHandler(PoshError::POSH__RUNTIME_LEADING_SLASH_PROVIDED, ErrorLevel::FATAL);
+        LogFatal() << "Cannot initialize runtime. The application name \"" << **name << "\" is not a valid file name.";
```
Comment (mossmaurice) [inline iceoryx_posh/source/runtime/posh_runtime.cpp:90]: I think we should give the user more info on what is allowed and what not. Where did you derive `isValidFileName` from? Is this defined in some POSIX / UNIX standard?

```suggestion
        LogFatal() << "Cannot initialize runtime. The application name \"" << **name << "\" is not a valid file name according to POSIX blublablubb";
```

File: iceoryx_posh/test/moduletests/test_posh_runtime.cpp [hunk 2]
```diff
@@ -148,12 +148,23 @@ TEST_F(PoshRuntime_test, NoAppName)
...
 
-    EXPECT_DEATH({ PoshRuntime::initRuntime(invalidAppName); }, "POSH__RUNTIME_LEADING_SLASH_PROVIDED");
+    for (auto i : {"/miau",
+                   "/fuu/bar",
+                   "plum/bus",
+                   ".",
+                   "..",
+                   "strawberriesWithMayonnaiseIs/..",
+                   "ohLookADot.",
+                   "amIADirectory/"})
```
Comment (mossmaurice) [inline iceoryx_posh/test/moduletests/test_posh_runtime.cpp:162]: I think you missed `IAmASpace ` and the empty case. In the end we're just duplicating the tests from `test_cxx_helplets.cpp`. Maybe this is not needed and we just test with one case because we know that the underlying `isValidFileName` works as expected?

File: iceoryx_posh/source/runtime/posh_runtime.cpp (hunk 1 above)
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/posh_runtime.cpp:90]: Please also update doxygen of `initRuntime` and add there how an valid name looks like e.g. `defined according to POSIX standard blublablubb`

**Review by mossmaurice — CHANGES_REQUESTED** [review]
Thanks for the PR! I left two remarks.

**Review by dkroenke — APPROVED** [review]

File: iceoryx_posh/test/moduletests/test_posh_runtime.cpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_posh/test/moduletests/test_posh_runtime.cpp]: @elfenpiff Ok. However, if we now support a new file system or platform we have to adapt two tests at the same time and one might forget either this one here or the on in helplets. Maybe we can share the test cases? It's not a biggie and can be solved on another PR.

**Review by mossmaurice — APPROVED** [review]
